<a href="https://colab.research.google.com/github/Nastya2006Fed/python-ai-Nastya-Fedorova/blob/main/notebooks/week2b_read_csv.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 📊 Week 2: Data Analysis — Чтение и проверка данных

**Цель**: Научиться читать CSV-файлы из репозитория GitHub в Google Colab и выполнять базовую проверку данных с помощью pandas.

**Данные:**
- **[`rivers.csv`](https://github.com/Nastya2006Fed/python-ai-Nastya-Fedorova/blob/main/data/rivers.csv)** — информация о реках: название, страна, длина, исток и устье

**Что мы делаем:**
1. Клонируем репозиторий GitHub в Colab
2. Читаем CSV-файлы в pandas DataFrame
3. Очищаем и переименовываем столбцы
4. Смотрим структуру данных и делаем быструю валидацию

## 🐱 [1] Клонируем репозиторий курса в Colab

In [29]:
# 🐱 Шаг 1. Клонируем репозиторий курса в Colab

import os

# 🔁 ИЗМЕНЕНО: имя репозитория и URL
repo = "python-ai-Nastya-Fedorova"  # было: "python-ai-template"
repo_path = f"/content/{repo}"

if not os.path.exists(repo_path):
    # 🔁 ИЗМЕНЕНО: URL вашего репозитория
    !git clone -q https://github.com/Nastya2006Fed/python-ai-Nastya-Fedorova.git

if os.getcwd() != repo_path:
    %cd {repo_path}

print("✅ Репозиторий готов, теперь мы работаем внутри папки", repo)

✅ Репозиторий готов, теперь мы работаем внутри папки python-ai-Nastya-Fedorova


## 📥 [2A] Простое чтение CSV-файлов в pandas

Сначала просто прочитаем оба CSV-файла в объекты `DataFrame`, без каких‑либо изменений.

После этого мы узнаем, сколько строк загружено в каждый датасет.

In [30]:
# 🐱 Шаг 2A. Чтение CSV-файлов в pandas

import pandas as pd

df_rivers = pd.read_csv("data/rivers.csv")

print("✅ Загружено строк в df_genre:", len(df_rivers))


✅ Загружено строк в df_genre: 563


## 🧹 [2B] Очистка и переименование столбцов

В вашем CSV-файле `rivers.csv` есть **технические столбцы**, которые полезны для Викиданных, но мешают простому анализу:

- Столбец `river` с URL (ссылкой на объект Wikidata) — **удаляем**, так как для базового анализа он не требуется.
- Столбцы `riverLabel`, `countryLabel`, `sourceLabel`, `mouthLabel` содержат читаемые подписи (названия) — **убираем суффикс `Label`** для удобства.

В этом шаге мы:
- 🔁 **удаляем столбец `river` с URL Wikidata**;
- 🔁 **переименовываем**: `riverLabel → river`, `countryLabel → country`, `sourceLabel → source`, `mouthLabel → mouth`;
- 🔁 **приводим столбец `length` к числовому типу** (`int`).

При приведении к числам мы используем:
- `pd.to_numeric(..., errors="coerce")` — преобразует значения в числа, некорректные значения превращает в `NaN`;
- `fillna(0)` — заменяет пропущенные значения (`NaN`) на 0;
- `astype(int)` — переводит столбец к целочисленному типу.

> ⚠️ **Важно:** если в ваших данных есть столбцы с URL Wikidata и столбцы вида `*Label`, этот шаг обязателен, чтобы получить аккуратные таблички для анализа. Столбец с URL можно сохранить, если нужно будет быстро перейти к оригинальной записи в Викиданных.

In [31]:
# 🧹 Шаг 2B. Очистка и переименование столбцов для rivers.csv

from IPython.display import display  # 🔁 ДОБАВЛЕНО: для красивого вывода таблиц

if "riverLabel" in df_rivers.columns:                             # Проверка: нужна ли очистка?

    # 🔁 ИСПРАВЛЕНО: не удаляем, а переименовываем столбец с URL Wikidata
    # Было: df_rivers = df_rivers.drop(columns=["URL"])  # ❌ ошибка: столбец ещё называется "river"
    # Стало:
    if "river" in df_rivers.columns and df_rivers["river"].dtype == "object" and df_rivers["river"].str.contains("wikidata", case=False, na=False).any():
        df_rivers = df_rivers.rename(columns={"river": "URL"})  # ✅ переименовываем, а не удаляем
        print("   → Столбец 'river' (URL) переименован в 'URL'")

    # Переименовываем столбцы, убирая суффикс Label
    df_rivers = df_rivers.rename(columns={
        "riverLabel": "river",          # ← название реки
        "countryLabel": "country",      # ← страна
        "sourceLabel": "source",        # ← исток
        "mouthLabel": "mouth",          # ← устье
    })

    # Приводим длину реки к числовому типу
    df_rivers["length"] = pd.to_numeric(
        df_rivers["length"], errors="coerce"
    ).fillna(0).astype(int)

    print("✅ df_rivers очищен и переименован")
else:
    print("⏭️ df_rivers уже очищен, пропускаем")

# 🔁 ИСПРАВЛЕНО: вывод таблицы без print() для HTML-форматирования
print("\n📋 Столбцы после очистки:", df_rivers.columns.tolist())
display(df_rivers.head())  # ✅ Красивая таблица вместо текста

print("\n✅ Данные готовы к анализу")

   → Столбец 'river' (URL) переименован в 'URL'
✅ df_rivers очищен и переименован

📋 Столбцы после очистки: ['URL', 'river', 'country', 'length', 'source', 'mouth']


,URL,river,country,length,source,mouth
0,http://www.wikidata.org/entity/Q272074,Снейк,США,1735,Йеллоустонский национальный парк,Колумбия
1,http://www.wikidata.org/entity/Q1279982,Куараи,Уругвай,351,Бразилия,Уругвай
2,http://www.wikidata.org/entity/Q1279982,Куараи,Бразилия,351,Бразилия,Уругвай
3,http://www.wikidata.org/entity/Q784682,Сож,Россия,648,Россия,Днепр
4,http://www.wikidata.org/entity/Q1279982,Куараи,Уругвай,351,Бразилия,Уругвай



✅ Данные готовы к анализу


## 🔢 [2C] Учёт уникальных рек: почему строк больше, чем рек?

**Проблема**: В датасете **563 строки**, но уникальных рек **меньше**. Почему?

Потому что одна река может протекать через несколько стран, и Викиданные добавляют **по одной строке на каждую страну**. Такой формат называется **«длинный формат»** (long format).

**Примеры из данных**:
| Река | URL (уникальный ID) | Страны в данных |
|------|---------------------|-----------------|
| Куараи | `.../Q1279982` | Уругвай, Бразилия, Уругвай (3 строки) |
| Нил | `.../Q3392` | Египет, Эфиопия (2 строки) |
| Снейк | `.../Q272074` | США (1 строка) |

**Зачем это важно**:
- 📉 Если считать среднюю длину по всем 563 строкам, то Нил (6650 км) будет учтён дважды, Амазонка — трижды → **среднее завысится**.
- 🌍 Если считать топ стран по числу рек, то трансграничные реки искусственно увеличат счётчик.
- 🔍 Для корректного анализа нужно работать с **уникальными реками**, а не со строками.

**Что делаем в этом шаге**:
1. Сравниваем: `len(df_rivers)` (строки) vs `df_rivers["URL"].nunique()` (уникальные реки).
2. Создаём `df_unique` — таблицу, где каждая река встречается **ровно один раз** (первое вхождение по URL).
3. Пересчитываем статистику (`describe()`, `value_counts()`) по `df_unique`.
4. 🔁 Выводим **оба варианта рядом**, чтобы увидеть разницу.

> 💡 **Аналогия**: если при подсчёте среднего роста в классе записать Васю в список три раза — результат исказится. Мы «оставляем Васю один раз».

**Ожидаемый результат** (пример):



In [32]:
# 🔢 Шаг 2C. Учёт уникальных рек по URL

print("🔍 Анализ дубликатов: строки vs уникальные реки\n")

# 1) Сравниваем количество строк и уникальных рек
total_rows = len(df_rivers)
unique_rivers = df_rivers["URL"].nunique()
diff = total_rows - unique_rivers

print(f"📊 Всего строк в df_rivers: {total_rows}")
print(f"🆔 Уникальных рек (по URL): {unique_rivers}")
print(f"📉 «Лишних» строк из-за трансграничных рек: {diff}")
print(f"   → В среднем каждая река представлена {total_rows/unique_rivers:.2f} раз\n")

# 2) Показываем примеры рек, которые встречаются в нескольких странах
print("🌍 Примеры трансграничных рек (одна река → несколько стран):")
multi_country = df_rivers.groupby("URL").filter(lambda x: x["country"].nunique() > 1)
if not multi_country.empty:
    examples = multi_country.groupby("URL")[["river", "country", "length"]].first().head(5)
    # Добавляем список стран для наглядности
    countries_list = multi_country.groupby("URL")["country"].apply(lambda x: ", ".join(x.unique()))
    examples["countries"] = countries_list
    display(examples[["river", "countries", "length"]])
else:
    print("   ⚠️ Не найдено рек, протекающих через несколько стран")

print("\n" + "-"*60 + "\n")

# 3) Создаём df_unique: одна строка на реку (первое вхождение по URL)
df_unique = df_rivers.drop_duplicates(subset="URL", keep="first")
print(f"✅ Создан df_unique: {len(df_unique)} строк (было {total_rows})")

# 4) Сравниваем статистику по длине: все строки vs уникальные реки
print("\n📏 Сравнение статистики по длине рек (км):")
stats_comparison = pd.DataFrame({
    "По всем строкам (df_rivers)": df_rivers["length"].describe(),
    "По уникальным рекам (df_unique)": df_unique["length"].describe()
}).round(1)
display(stats_comparison)

# 5) Сравниваем топ стран: до и после удаления дубликатов
print("\n🌍 Топ-10 стран по числу рек (сравнение):")
country_comparison = pd.DataFrame({
    "Все строки": df_rivers["country"].value_counts().head(10),
    "Уникальные реки": df_unique["country"].value_counts().head(10)
})
country_comparison["Разница"] = country_comparison["Все строки"] - country_comparison["Уникальные реки"]
display(country_comparison)

# 6) Краткий вывод
print("\n💡 Вывод:")
if stats_comparison.loc["mean", "По всем строкам (df_rivers)"] > stats_comparison.loc["mean", "По уникальным рекам (df_unique)"]:
    print("   ⚠️ Средняя длина по всем строкам ЗАВЫШЕНА из-за повторного учёта длинных трансграничных рек.")
    print("   ✅ Для корректного анализа дальше используем df_unique.")
else:
    print("   ✅ Разница в средней длине незначительна, но df_unique всё равно предпочтительнее для чистоты анализа.")

# 7) Сохраняем df_unique для дальнейшего использования
# (теперь в следующих шагах работаем с df_unique, а не df_rivers)
print(f"\n🎯 Готово! Дальше используем df_unique ({len(df_unique)} уникальных рек)")

🔍 Анализ дубликатов: строки vs уникальные реки

📊 Всего строк в df_rivers: 563
🆔 Уникальных рек (по URL): 399
📉 «Лишних» строк из-за трансграничных рек: 164
   → В среднем каждая река представлена 1.41 раз

🌍 Примеры трансграничных рек (одна река → несколько стран):


,river,countries,length
URL,,,
http://www.wikidata.org/entity/Q104437,Юкон,"Канада, США",3190
http://www.wikidata.org/entity/Q118251,Мадейра,"Бразилия, Боливия",1450
http://www.wikidata.org/entity/Q1265,Колорадо,"США, Мексика",2334
http://www.wikidata.org/entity/Q127892,Парана,"Бразилия, Парагвай",4880
http://www.wikidata.org/entity/Q1279982,Куараи,"Уругвай, Бразилия",351



------------------------------------------------------------

✅ Создан df_unique: 399 строк (было 563)

📏 Сравнение статистики по длине рек (км):


,По всем строкам (df_rivers),По уникальным рекам (df_unique)
count,563.0,399.0
mean,1149.5,922.6
std,1215.5,913.1
min,301.0,301.0
25%,455.0,437.0
50%,686.0,600.0
75%,1256.5,1027.0
max,6650.0,6650.0



🌍 Топ-10 стран по числу рек (сравнение):


,Все строки,Уникальные реки,Разница
country,,,
Австралия,10.0,9.0,1.0
Бразилия,27.0,19.0,8.0
Индия,25.0,17.0,8.0
Канада,85.0,76.0,9.0
Китай,48.0,35.0,13.0
Мексика,NaN,7.0,NaN
Перу,9.0,7.0,2.0
Россия,152.0,127.0,25.0
США,128.0,70.0,58.0



💡 Вывод:
   ⚠️ Средняя длина по всем строкам ЗАВЫШЕНА из-за повторного учёта длинных трансграничных рек.
   ✅ Для корректного анализа дальше используем df_unique.

🎯 Готово! Дальше используем df_unique (399 уникальных рек)


## 🔍 [2D] Проверка заполненности: исток (`source`) и устье (`mouth`)

**Проблема**: В Викиданных данные часто бывают неполными. У части рек исток или устье могут быть не указаны — такие ячейки в pandas отображаются как `NaN` (Not a Number).

**Зачем это проверять**:
- 📉 Если устье заполнено только у 40% рек, то график «Куда текут реки» покажет **лишь часть картины**, а не все реки.
- ⚠️ Делать выводы по неполным данным без предупреждения — это как составлять рейтинг студентов, если у половины оценка не выставлена.
- ✅ Честное указание на пробелы в данных повышает доверие к анализу.

**Что делаем в этом шаге**:
1. Считаем долю непустых значений для `source` и `mouth` (в процентах).
2. Выводим результат с округлением до 1 знака после запятой.
3. 📝 Формулируем вывод: какие поля заполнены хорошо (>80%), удовлетворительно (50–80%), плохо (<50%).
4. 🔁 Даём рекомендации: как учитывать пробелы при построении графиков.

**Пример вывода**:

In [33]:
# 🔍 Шаг 2D. Проверка заполненности столбцов source и mouth

print("📋 Проверка качества данных: заполненность столбцов\n")

# 1) Считаем долю непустых значений (в процентах)
source_fill_rate = df_unique["source"].notna().mean() * 100
mouth_fill_rate = df_unique["mouth"].notna().mean() * 100

# Округляем до 1 знака после запятой
source_fill_rate = round(source_fill_rate, 1)
mouth_fill_rate = round(mouth_fill_rate, 1)

print(f"✅ Исток (source): заполнен на {source_fill_rate}%")
print(f"✅ Устье (mouth): заполнен на {mouth_fill_rate}%")

# 2) Оцениваем качество заполнения
def rate_quality(fill_rate):
    if fill_rate >= 80:
        return "🟢 Отлично", "можно использовать без оговорок"
    elif fill_rate >= 50:
        return "🟡 Удовлетворительно", "можно использовать с пометкой «частичные данные»"
    else:
        return "🔴 Плохо", "визуализация покажет лишь часть картины — требуется осторожность"

source_quality, source_note = rate_quality(source_fill_rate)
mouth_quality, mouth_note = rate_quality(mouth_fill_rate)

print(f"\n🎯 Оценка качества:")
print(f"   source: {source_quality} — {source_note}")
print(f"   mouth:  {mouth_quality} — {mouth_note}")

# 3) Показываем количество NaN для контекста
total_rivers = len(df_unique)
source_nan = df_unique["source"].isna().sum()
mouth_nan = df_unique["mouth"].isna().sum()

print(f"\n📊 Контекст (на основе {total_rivers} уникальных рек):")
print(f"   source: {source_nan} пропусков ({round(source_nan/total_rivers*100, 1)}%)")
print(f"   mouth:  {mouth_nan} пропусков ({round(mouth_nan/total_rivers*100, 1)}%)")

# 4) Рекомендации для следующего шага (визуализация)
print(f"\n💡 Рекомендации для визуализации:")
if mouth_fill_rate < 50:
    print("   ⚠️ При построении графика «Устья рек» обязательно укажите:")
    print(f"      «Показаны данные для {mouth_fill_rate}% рек. Остальные устья не указаны в источнике.»")
if source_fill_rate < 50:
    print("   ⚠️ При анализе истоков добавьте аналогичную пометку о неполноте данных.")

# 5) Опционально: сохранить метрики для отчёта
# fill_rates = {"source": source_fill_rate, "mouth": mouth_fill_rate}
# print(f"\n💾 Метрики заполненности: {fill_rates}")

print(f"\n✅ Проверка завершена. Готовы к визуализации с учётом качества данных.")

📋 Проверка качества данных: заполненность столбцов

✅ Исток (source): заполнен на 100.0%
✅ Устье (mouth): заполнен на 100.0%

🎯 Оценка качества:
   source: 🟢 Отлично — можно использовать без оговорок
   mouth:  🟢 Отлично — можно использовать без оговорок

📊 Контекст (на основе 399 уникальных рек):
   source: 0 пропусков (0.0%)
   mouth:  0 пропусков (0.0%)

💡 Рекомендации для визуализации:

✅ Проверка завершена. Готовы к визуализации с учётом качества данных.


## 🔍 [3] Обзор данных: структура и первые строки

Сделаем короткий обзор DataFrame с данными о реках:

- посмотрим размер таблицы (`shape`);
- выведем список столбцов;
- посмотрим первые несколько строк;
- 🔁 **дополнительно посчитаем базовую статистику по длине рек (`length`)**.

Для удобства напишем маленькую функцию `show_info(df, name)`, чтобы не повторять один и тот же код.

In [34]:
def show_info(df, name, n=5):
    """Краткий обзор DataFrame: имя, размер, список столбцов и первые строки."""
    print(f"\n📊 {name}")
    print("Размер:", df.shape)
    print("Столбцы:", ", ".join(df.columns))
    print("\nПервые строки:")
    display(df.head(n))  # 🔁 ИЗМЕНЕНО: display() вместо print() для красивой таблицы

    # 🔁 ИЗМЕНЕНО: статистика по length вместо capital_cost
    numeric_cols = df.select_dtypes(include=['number']).columns
    if len(numeric_cols) > 0:
        print("\n📈 Базовая статистика по числовым столбцам:")
        display(df[numeric_cols].describe())

# 🔍 Шаг 3. Обзор данных

# 🔁 ИЗМЕНЕНО: работаем с одним DataFrame df_rivers
show_info(df_rivers, "Реки: название, страна, длина, исток, устье (df_rivers)")


📊 Реки: название, страна, длина, исток, устье (df_rivers)
Размер: (563, 6)
Столбцы: URL, river, country, length, source, mouth

Первые строки:


,URL,river,country,length,source,mouth
0,http://www.wikidata.org/entity/Q272074,Снейк,США,1735,Йеллоустонский национальный парк,Колумбия
1,http://www.wikidata.org/entity/Q1279982,Куараи,Уругвай,351,Бразилия,Уругвай
2,http://www.wikidata.org/entity/Q1279982,Куараи,Бразилия,351,Бразилия,Уругвай
3,http://www.wikidata.org/entity/Q784682,Сож,Россия,648,Россия,Днепр
4,http://www.wikidata.org/entity/Q1279982,Куараи,Уругвай,351,Бразилия,Уругвай



📈 Базовая статистика по числовым столбцам:


,length
count,563.000000
mean,1149.477798
std,1215.502726
min,301.000000
25%,455.000000
50%,686.000000
75%,1256.500000
max,6650.000000


## 📏🌊 [4] Быстрая проверка и валидация данных: перцентили длины и устья-хабы

**Цель**: Провести два взаимосвязанных численных анализа без визуализации:
1. 📊 **Перцентили длины** — разделить реки на группы по длине и сравнить их статистику.
2. 🌊 **Устья-хабы** — выявить ключевые «узлы» речной системы по количеству и суммарной длине притоков.

---

### 📏 Часть A: Перцентили длины

**Что делаем**:
1. Рассчитываем перцентили: 25% (Q1), 50% (медиана), 75% (Q3).
2. Присваиваем каждой реке метку группы:
   - 🟢 «Короткие»: 0–25%
   - 🟡 «Средние»: 25–50%
   - 🟠 «Длинные»: 50–75%
   - 🔴 «Гиганты»: 75–100%
3. Для каждой группы считаем: количество рек, среднюю/мин/макс длину, топ-3 страны.

**Зачем**: Понять распределение рек по длине и выявить, характерны ли «гиганты» для определённых стран.

---

### 🌊 Часть B: Устья-хабы

**Что делаем**:
1. Группируем данные по столбцу `mouth` (устье).
2. Для каждого устья считаем:
   - количество впадающих рек;
   - суммарную и среднюю длину притоков;
   - список уникальных стран-источников.
3. Строим два рейтинга:
   - 🔢 Топ-15 устьев по **количеству** притоков;
   - 📏 Топ-15 устьев по **суммарной длине** притоков.
4. 🔍 Находим «международные хабы»: устья, куда впадают реки из ≥3 стран.

**Зачем**: Выявить географические закономерности и ключевые узлы гидрографической сети.

> 💡 Все результаты выводятся в виде таблиц и текста — визуализация будет на следующем этапе.

In [35]:
# 📏🌊 Шаг 4. Быстрая проверка и валидация данных: перцентили длины + устья-хабы

print("🚀 Запуск комплексного анализа речных данных\n")

# =============================================================================
# 📏 ЧАСТЬ A: Перцентили длины рек
# =============================================================================
print("📊 ЧАСТЬ A: Анализ длины рек по перцентилям\n")

# 1) Рассчитываем перцентили
q1 = df_rivers["length"].quantile(0.25)
q2 = df_rivers["length"].quantile(0.50)
q3 = df_rivers["length"].quantile(0.75)
max_len = df_rivers["length"].max()

print(f"📐 Границы перцентилей:")
print(f"   Q1 (25%): {q1:.0f} км | Q2 (медиана): {q2:.0f} км | Q3 (75%): {q3:.0f} км | Макс: {max_len:.0f} км\n")

# 2) Присваиваем метки групп
def assign_group(length):
    if length <= q1:
        return "🟢 Короткие (0-25%)"
    elif length <= q2:
        return "🟡 Средние (25-50%)"
    elif length <= q3:
        return "🟠 Длинные (50-75%)"
    else:
        return "🔴 Гиганты (75-100%)"

df_rivers["length_group"] = df_rivers["length"].apply(assign_group)

# 3) Сводная статистика по группам
group_stats = df_rivers.groupby("length_group").agg(
    count=("river", "count"),
    avg_length=("length", "mean"),
    min_length=("length", "min"),
    max_length=("length", "max"),
    total_length=("length", "sum")
).round(1)

group_order = ["🟢 Короткие (0-25%)", "🟡 Средние (25-50%)", "🟠 Длинные (50-75%)", "🔴 Гиганты (75-100%)"]
group_stats = group_stats.reindex(group_order)

print("📋 Статистика по группам длины:")
display(group_stats)

# 4) Топ-3 страны внутри каждой группы
print("\n🌍 Топ-3 страны в каждой группе по числу рек:")
for group in group_order:
    subset = df_rivers[df_rivers["length_group"] == group]
    top_countries = subset["country"].value_counts().head(3)
    print(f"\n{group}:")
    for country, count in top_countries.items():
        print(f"   • {country}: {count} рек")

# 5) Примеры рек-гигантов
print("\n🔴 Топ-10 рек из группы «Гиганты»:")
giants = df_rivers[df_rivers["length_group"] == "🔴 Гиганты (75-100%)"].nlargest(10, "length")[["river", "country", "length", "mouth"]]
display(giants)

print("\n" + "="*70 + "\n")

# =============================================================================
# 🌊 ЧАСТЬ B: Устья-хабы
# =============================================================================
print("🌊 ЧАСТЬ B: Анализ устьев как речных хабов\n")

# 1) Базовая агрегация по устьям
mouth_stats = df_rivers.groupby("mouth").agg(
    river_count=("river", "count"),
    total_length=("length", "sum"),
    avg_length=("length", "mean"),
    countries=("country", lambda x: ", ".join(sorted(x.unique())))
).round(1)

mouth_stats["unique_countries"] = df_rivers.groupby("mouth")["country"].nunique()
mouth_stats = mouth_stats.reset_index()

# 2) Топ-15 устьев по количеству притоков
print("🔢 Топ-15 устьев по числу впадающих рек:")
top_by_count = mouth_stats.nlargest(15, "river_count")[["mouth", "river_count", "total_length", "unique_countries"]]
display(top_by_count)

# 3) Топ-15 устьев по суммарной длине притоков
print("\n📏 Топ-15 устьев по суммарной длине притоков (км):")
top_by_length = mouth_stats.nlargest(15, "total_length")[["mouth", "river_count", "total_length", "avg_length", "unique_countries"]]
display(top_by_length)

# 4) Международные хабы: устья с реками из ≥3 стран
print("\n🌍 Международные устья-хабы (реки из ≥3 стран):")
international = mouth_stats[mouth_stats["unique_countries"] >= 3].sort_values("unique_countries", ascending=False)
if len(international) > 0:
    display(international[["mouth", "river_count", "total_length", "unique_countries", "countries"]].head(15))
else:
    print("   ⚠️ Не найдено устьев с реками из 3+ стран")

# 5) Детали для топ-3 устьев по суммарной длине
print("\n🔎 Детали для топ-3 устьев по суммарной длине:")
top3_mouths = mouth_stats.nlargest(3, "total_length")["mouth"]
for mouth in top3_mouths:
    subset = df_rivers[df_rivers["mouth"] == mouth]
    print(f"\n📍 Устье: {mouth}")
    print(f"   Рек: {len(subset)} | Суммарная длина: {subset['length'].sum():.0f} км")
    print(f"   Страны: {', '.join(subset['country'].unique())}")
    print(f"   Примеры рек: {', '.join(subset['river'].head(5))}")

# 🔁 Опционально: убрать временный столбец
# df_rivers = df_rivers.drop(columns=["length_group"])

print("\n" + "="*70)
print("✅ Комплексный анализ завершён: перцентили + устья-хабы")

🚀 Запуск комплексного анализа речных данных

📊 ЧАСТЬ A: Анализ длины рек по перцентилям

📐 Границы перцентилей:
   Q1 (25%): 455 км | Q2 (медиана): 686 км | Q3 (75%): 1256 км | Макс: 6650 км

📋 Статистика по группам длины:


,count,avg_length,min_length,max_length,total_length
length_group,,,,,
🟢 Короткие (0-25%),141,376.6,301,454,53105
🟡 Средние (25-50%),141,553.2,456,686,78003
🟠 Длинные (50-75%),140,916.3,688,1242,128277
🔴 Гиганты (75-100%),141,2750.1,1271,6650,387771



🌍 Топ-3 страны в каждой группе по числу рек:

🟢 Короткие (0-25%):
   • США: 43 рек
   • Россия: 32 рек
   • Канада: 30 рек

🟡 Средние (25-50%):
   • Россия: 45 рек
   • США: 37 рек
   • Канада: 23 рек

🟠 Длинные (50-75%):
   • Россия: 44 рек
   • США: 33 рек
   • Канада: 23 рек

🔴 Гиганты (75-100%):
   • Россия: 31 рек
   • Китай: 23 рек
   • Уругвай: 18 рек

🔴 Топ-10 рек из группы «Гиганты»:


,river,country,length,mouth
226,Нил,Египет,6650,Средиземное море
227,Нил,Эфиопия,6650,Средиземное море
389,Нил,Египет,6650,Средиземное море
390,Нил,Эфиопия,6650,Средиземное море
111,Амазонка,Бразилия,6400,Атлантический океан
112,Амазонка,Перу,6400,Атлантический океан
113,Амазонка,Колумбия,6400,Атлантический океан
126,Амазонка,Бразилия,6400,Атлантический океан
127,Амазонка,Перу,6400,Атлантический океан
128,Амазонка,Колумбия,6400,Атлантический океан




🌊 ЧАСТЬ B: Анализ устьев как речных хабов

🔢 Топ-15 устьев по числу впадающих рек:


,mouth,river_count,total_length,unique_countries
117,Ла-Плата,26,65080,3
126,Миссури,23,14254,2
37,Атлантический океан,19,54431,9
139,Обь,11,18126,2
78,Енисей,10,9790,1
125,Миссисипи,9,11078,1
134,Нижняя Тунгуска,9,4313,1
192,Тихий океан,9,7187,6
200,Уругвай,8,3126,2
28,Амазонка,7,9060,4



📏 Топ-15 устьев по суммарной длине притоков (км):


,mouth,river_count,total_length,avg_length,unique_countries
117,Ла-Плата,26,65080,2503.1,3
37,Атлантический океан,19,54431,2864.8,9
186,Средиземное море,6,27445,4574.2,3
139,Обь,11,18126,1647.8,2
104,Карское море,5,14656,2931.2,1
126,Миссури,23,14254,619.7,2
33,Аравийское море,5,14035,2807.0,2
60,Восточно-Китайское море,2,12600,6300.0,1
219,Южно-Китайское море,4,12249,3062.2,1
62,Ганг,7,11337,1619.6,2



🌍 Международные устья-хабы (реки из ≥3 стран):


,mouth,river_count,total_length,unique_countries,countries
37,Атлантический океан,19,54431,9,"Бразилия, Гамбия, Колумбия, Марокко, Перу, США..."
192,Тихий океан,9,7187,6,"Канада, Колумбия, Мексика, Перу, США, Чили"
28,Амазонка,7,9060,4,"Боливия, Бразилия, Перу, Эквадор"
50,Берингово море,5,9790,3,"Канада, Россия, США"
117,Ла-Плата,26,65080,3,"Бразилия, Парагвай, Уругвай"
186,Средиземное море,6,27445,3,"Египет, Марокко, Эфиопия"



🔎 Детали для топ-3 устьев по суммарной длине:

📍 Устье: Ла-Плата
   Рек: 26 | Суммарная длина: 65080 км
   Страны: Бразилия, Парагвай, Уругвай
   Примеры рек: Парана, Парана, Парана, Парана, Уругвай

📍 Устье: Атлантический океан
   Рек: 19 | Суммарная длина: 54431 км
   Страны: ЮАР, Бразилия, Перу, Колумбия, Марокко, Гамбия, Сенегал, Суринам, США
   Примеры рек: Оранжевая, Амазонка, Амазонка, Амазонка, Риу-Доси

📍 Устье: Средиземное море
   Рек: 6 | Суммарная длина: 27445 км
   Страны: Египет, Марокко, Эфиопия
   Примеры рек: Рашид, Мулуя, Нил, Нил, Нил

✅ Комплексный анализ завершён: перцентили + устья-хабы


## 📝 Summary

**Что мы сделали в этом ноутбуке (Week 2):**

- ✅ Клонировали репозиторий GitHub в Colab
- ✅ Прочитали 2 CSV-файла из `data/examples/`
- ✅ Удалили URL Wikidata и переименовали столбцы (`*Label → короткие имена`)
- ✅ Проверили структуру данных (размер, столбцы, первые строки)
- ✅ Выполнили быструю валидацию:
  - количество уникальных фильмов, стран, жанров
  - диапазоны значений
  - топ стран и жанров по числу записей
  - типы оценок и результатов

Теперь у нас есть **аккуратные, проверенные таблицы**, с которыми удобно работать дальше.

В отдельном ноутбуке для следующей недели мы будем использовать **те же данные** для:
- более сложного анализа (группировки, фильтрация),
- и построения визуализаций (графики и диаграммы). 🎨